### PR TITLE
fix bug with openssl when cmake is missing

### LINKED ISSUE #11145: problem with dependency('openssl')
**Describe the bug**

An internal error, probably due to some refactoring. Keyword: `log_tried()`
I don't think the specifics of the system setup are that relevant.

**To Reproduce**

On Windows, inside an MSYS2 container, without CMake installed. This `meson.build` is the only required file:

```
project('bug', 'cpp')
openssl_dep = dependency('openssl')
```

Output of `meson setup build`
```
The Meson build system
Version: 0.64.1
Source dir: C:/MSYS2/home/hmoebius/projects/mesontest
Build dir: C:/MSYS2/home/hmoebius/projects/mesontest/build
Build type: native build
Project name: bug
Project version: undefined
C++ compiler for the host machine: cl (msvc 19.29.30146 "Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30146 for x64")
C++ linker for the host machine: link link 14.29.30146.0
Host machine cpu family: x86_64
Host machine cpu: x86_64
Did not find pkg-config by name 'pkg-config'
Found Pkg-config: NO
Did not find CMake 'cmake'
Found CMake: NO
Traceback (most recent call last):
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/dependencies/detect.py", line 112, in find_external_dependency
    d = c()
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/dependencies/misc.py", line 702, in <lambda>
    cmake_class=lambda name, env, kwargs: CMakeDependency('OpenSSL', env, dict(kwargs, modules=['OpenSSL::Crypto', 'OpenSSL::SSL'])),
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/dependencies/cmake.py", line 127, in __init__
    raise DependencyException(msg)
mesonbuild.dependencies.base.DependencyException: CMake binary for machine 1 not found. Giving up.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/mesonmain.py", line 194, in run
    return options.run_func(options)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/msetup.py", line 307, in run
    app.generate()
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/msetup.py", line 186, in generate
    self._generate(env)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/msetup.py", line 230, in _generate
    intr.run()
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreter/interpreter.py", line 2895, in run
    super().run()
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 149, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 174, in evaluate_codeblock
    raise e
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 167, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 182, in evaluate_statement
    self.assignment(cur)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 575, in assignment
    value = self.evaluate_statement(node.value)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 180, in evaluate_statement
    return self.function_call(cur)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 462, in function_call
    res = func(node, func_args, kwargs)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/decorators.py", line 775, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/decorators.py", line 775, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/decorators.py", line 775, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  [Previous line repeated 5 more times]
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/decorators.py", line 109, in wrapped
    ret = f(*wrapped_args, **wrapped_kwargs)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/decorators.py", line 127, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreterbase/decorators.py", line 277, in wrapper
    return f(*nargs, **wrapped_kwargs)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreter/interpreter.py", line 1701, in func_dependency
    d = df.lookup(kwargs)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreter/dependencyfallbacks.py", line 354, in lookup
    dep = func(kwargs, func_args, func_kwargs)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/interpreter/dependencyfallbacks.py", line 87, in _do_dependency
    dep = dependencies.find_external_dependency(name, self.environment, kwargs)
  File "C:/MSYS2/mingw64/lib/python3.10/site-packages/mesonbuild/dependencies/detect.py", line 117, in find_external_dependency
    bettermsg = f'Dependency lookup for {name} with method {c.func.log_tried()!r} failed: {e}'
AttributeError: 'function' object has no attribute 'log_tried'

meson.build:2:0: ERROR: Unhandled python exception

    This is a Meson bug and should be reported!
```

**Expected behavior**

No unhandled exception. :)

**system parameters**
* native build
* Windows build, inside MSYS2 container
* No CMake installed
* Python 3.10.8
* Meson 0.64.1
* Ninja 1.11.1

### LINKED ISSUE #12098: Unhandled python exception
**Describe the bug**

```
Did not find pkg-config by name 'pkg-config'
Found Pkg-config: NO
Did not find CMake 'cmake'
Found CMake: NO
Traceback (most recent call last):
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/dependencies/detect.py", line 124, in find_external_dependency
    d = c()
        ^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/dependencies/misc.py", line 602, in <lambda>
    cmake_class=lambda name, env, kwargs: CMakeDependency('OpenSSL', env, dict(kwargs, modules=['OpenSSL::Crypto', 'OpenSSL::SSL'])),
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/dependencies/cmake.py", line 127, in __init__
    raise DependencyException(msg)
mesonbuild.dependencies.base.DependencyException: CMake binary for machine 1 not found. Giving up.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/mesonmain.py", line 194, in run
    return options.run_func(options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/msetup.py", line 358, in run
    app.generate()
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/msetup.py", line 183, in generate
    return self._generate(env, capture, vslite_ctx)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/msetup.py", line 228, in _generate
    intr.run()
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreter/interpreter.py", line 2997, in run
    super().run()
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 164, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 190, in evaluate_codeblock
    raise e
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 182, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 198, in evaluate_statement
    self.assignment(cur)
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 635, in assignment
    value = self.evaluate_statement(node.value)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 196, in evaluate_statement
    return self.function_call(cur)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 521, in function_call
    res = func(node, func_args, kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/decorators.py", line 826, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/decorators.py", line 826, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/decorators.py", line 826, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  [Previous line repeated 4 more times]
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/decorators.py", line 109, in wrapped
    ret = f(*wrapped_args, **wrapped_kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/decorators.py", line 127, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/decorators.py", line 277, in wrapper
    return f(*nargs, **wrapped_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreterbase/decorators.py", line 596, in wrapper
    return f(*wrapped_args, **wrapped_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreter/interpreter.py", line 1774, in func_dependency
    d = df.lookup(kwargs)
        ^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreter/dependencyfallbacks.py", line 355, in lookup
    dep = func(kwargs, func_args, func_kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/interpreter/dependencyfallbacks.py", line 87, in _do_dependency
    dep = dependencies.find_external_dependency(name, self.environment, kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:/MSYS2/mingw64/lib/python3.11/site-packages/mesonbuild/dependencies/detect.py", line 129, in find_external_dependency
    bettermsg = f'Dependency lookup for {name} with method {c.func.log_tried()!r} failed: {e}'
                                                            ^^^^^^^^^^^^^^^^
AttributeError: 'function' object has no attribute 'log_tried'

meson.build:15:21: ERROR: Unhandled python exception

    This is a Meson bug and should be reported!
```

**To Reproduce**

The problem pops up while checking for the openssl dependency. Neither pkg-config, nor cmake are available.

A meson.build as simple as this is enough, if pkg-config or cmake are not available.

```
project('bug', 'cpp')
openssl_dep = dependency('openssl')
```

**Expected behavior**

A proper error message would be nice.

**system parameters**
* Native build
* Windows, inside an MSYS2
* python 3.11.4
* meson 1.2.1
* ninja 1.11.1

### PR DESCRIPTION
**Pull request by bruchar1**
Fixes #12098
Fixes #11145

DependencyFactory was returning a lambda, but it has no log_tried() function.

### CODE REVIEW

**Comment by tristan957** [comment]
Do you think we can test this in CI or not?

File: mesonbuild/dependencies/cmake.py [hunk 1]
```diff
@@ -652,3 +652,19 @@ def get_variable(self, *, cmake: T.Optional[str] = None, pkgconfig: T.Optional[s
         if default_value is not None:
             return default_value
         raise DependencyException(f'Could not get cmake variable and no default provided for {self!r}')
+
+
+class CMakeDependencyFactory:
+
+    def __init__(self, name: T.Optional[str] = None, modules: T.Optional[T.List[str]] = None):
```
Comment (tristan957) [inline mesonbuild/dependencies/cmake.py:659]: Seems like a better API might be `modules: T.List[str] = []`.

Comment (tristan957) [reply mesonbuild/dependencies/cmake.py:659]: Are there even any users of this API where modules isn't passed?

Comment (QuLogic) [reply mesonbuild/dependencies/cmake.py:659]: One must be careful of using mutable objects as default arguments; the _defaults_ change if you modify them in the function. That doesn't seem to happen here, but it's still considered an anti-pattern to use them.

**Comment by bruchar1** [comment]
> Do you think we can test this in CI or not?

I'm not sure it would be easy. I requires openssl to be not available, and cmake neither.

**Comment by eli-schwartz** [comment]
Maybe try searching with `method: 'cmake'` plus whatever cmake settings are necessary to make it *think* it can't find openssl, e.g.searching only in self.builddir.

**Comment by bruchar1** [comment]
> Maybe try searching with `method: 'cmake'` plus whatever cmake settings are necessary to make it _think_ it can't find openssl, e.g.searching only in self.builddir.

I succeed writing a testcase that triggers the bug on the master branch, by using a native file with a dummy cmake path. But... in the fix branch, it triggers a DependencyException from dependencies/cmake.py:127

**Comment by bruchar1** [comment]
Ok, I think I succeed writing it as a unittest.

**Comment by bruchar1** [comment]
> Maybe it should be in platform agnostic tests instead?

Yes. I wasn't sure where to put it. Moved.

**Comment by bruchar1** [comment]
Fixes #11145

**Comment by bruchar1** [comment]
A fix was previously proposed in #11147.

**Pull request merged**